multiprofit/fit_psf.py

```diff
@@ -0,0 +1,271 @@
+            config = CatalogPsfFitterConfig()
+        if logger is None:
+            logger = CatalogPsfFitter._get_logger()
+
+        if len(self.errors_expected) != len(config.flag_errors):
+            raise ValueError(f"{self.errors_expected=} keys not same len as {config.flag_errors=}")
+        errors_bad = {}
+        errors_recast = {}
+        for error_name, error_type in self.errors_expected.items():
+            if error_type in errors_recast:
+                errors_bad[error_name] = error_type
+            else:
+                errors_recast[error_type] = error_name
+        if errors_bad:
+            raise ValueError(f"{self.errors_expected=} keys contain duplicates from {config.flag_errors=}")
+
+        model_source = make_psf_source(sigma_xs=config.sigmas)
+        n_gaussians = len(config.sigmas)
+        params = tuple({x: None for x in model_source.parameters([], g2f.ParamFilter(fixed=False))})
+        filter_flux = g2f.ParamFilter(nonlinear=False, channel=g2f.Channel.NONE)
+        # Make an ordered set
+        flux_total = tuple({x: None for x in model_source.parameters(paramfilter=filter_flux)}.keys())
+        if len(flux_total) != 1:
+            raise RuntimeError(f"len({flux_total=}) != 1; PSF model is badly-formed")
+        flux_total = flux_total[0]
+        filter_flux = g2f.ParamFilter(linear=False, channel=g2f.Channel.NONE, fixed=False)
+        # TODO: Remove isinstance when channel filtering is fixed
+        fluxfracs = tuple({x: None for x in model_source.parameters(paramfilter=filter_flux)
+                           if isinstance(x, g2f.ProperFractionParameterD)}.keys())
+        if len(fluxfracs) != (n_gaussians - 1):
+            raise RuntimeError(f"len({fluxfracs=}) != {(n_gaussians - 1)=}; PSF model is badly-formed")
+        gaussian = model_source.components[0]
+        cenx, ceny = gaussian.parameters()[:2]
+        if not (isinstance(cenx, g2f.CentroidXParameterD) and isinstance(ceny, g2f.CentroidYParameterD)):
+            raise RuntimeError(f"{cenx=}, {ceny=} have unexpected non-centroid types")
+        model_psf = make_psfmodel_null()
+
+        catalog = catexp.get_catalog()
+        n_rows = len(catalog)
+        range_idx = range(n_rows)
+
+        columns = config.schema()
+        keys = [column.key for column in columns]
+        prefix = config.prefix_column
+        idx_flag_first = keys.index("unknown_flag")
+        idx_var_first = keys.index("cen_x")
+        columns_write = [f"{prefix}{col.key}" for col in columns[idx_var_first:]]
+        dtypes = [(f'{prefix if col.key != "id" else ""}{col.key}', col.dtype) for col in columns]
+        meta = {'config': config.toDict()}
+        results = Table(data=np.full(n_rows, np.nan, dtype=dtypes), units=[x.unit for x in columns],
+                        meta=meta)
+        # Set nan-default flags to False instead
+        for flag in columns[idx_flag_first:idx_var_first]:
+            results[flag.key] = False
```

Comment:
The entire table is initialized to nan above, because I want the float quantities to stay nan if it fails, but not the flags. If there's a better way to do that, I'm all ears.

Reply:
Wow, I just tested the above and was surprised to find out `NaN` `bool` mapped to `True`, and I'm not happy about it.